src/StreamFactoryInterface.php

```diff
@@ -7,17 +7,32 @@
 interface StreamFactoryInterface
```

Comment:
When we were developing PSR-7, we were *definitely* considering that the code backing `StreamInterface` *might not be a PHP stream resource*. While that's the primary use case, we did envision using generators, callbacks, iterators, etc. as stream sources; this is commonly done in node specifically and JS more generally, and we saw it as a viable use case.

However, please note I used the phrase "primary use case"; the resource-based streams are definitely the principal use case, and what I'd expect any PSR-7 implementation to provide; any other types will be specific to the implementing library. (Diactoros provides only resource and callback streams currently, for instance.)